Running appliance-build manually without Jenkins fails when fetching linux-pkg packages

This was caused by the linux-pkg-rework.
Error is:
```
+ aws s3 sync --only-show-errors s3://snapshot-de-images/builds/jenkins-ops/devops-gate/master/linux-pkg/master/combine-packages/post-push/350/packages .
+ for subdir in */
+ pushd '*/'
FAILURE: Build failed with an exception.
```
The reason for the failure is that the `packages` directory is created by Jenkins when resolving and fetching all the component packages in the combined-packages bundle.

--- a/build.gradle
+++ b/build.gradle
@@ -53,8 +53,7 @@ for (variant in allVariants) {
 
         for (envVar in ["DELPHIX_PLATFORMS", 
                         "AWS_S3_URI_LIVEBUILD_ARTIFACTS",
-                        "AWS_S3_URI_UPGRADE_VERIFICATION",
-                        "AWS_S3_PREFIX_UPGRADE_VERIFICATION"]) {
+                        "AWS_S3_URI_COMBINED_PACKAGES"]) {
             inputs.property(envVar, System.getenv(envVar)).optional(true)
         }
 

--- a/live-build/build.gradle
+++ b/live-build/build.gradle
@@ -21,7 +21,7 @@ apply from: "${rootProject.projectDir}/gradle-lib/util.gradle"
 task ancillaryRepository(type: Exec) {
     inputs.file "${rootProject.projectDir}/scripts/build-ancillary-repository.sh"
 
-    for (envVar in ["COMBINED_PACKAGES_S3_URL"]) {
+    for (envVar in ["AWS_S3_URI_COMBINED_PACKAGES"]) {
         inputs.property(envVar, System.getenv(envVar)).optional(true)
     }
 

--- a/scripts/aptly-repo-from-debs.sh
+++ b/scripts/aptly-repo-from-debs.sh
@@ -61,18 +61,22 @@ for deb_tarball in "$LIVE_BUILD_OUTPUT_DIR/$APPLIANCE_VARIANT"*.debs.tar.gz; do
 done
 
 #
-# Download the delphix upgrade verification debian package.
-# Note, we always pull from the "master" build of the verification
-# package, no matter what the UPSTREAM_BRANCH of the appliance is that
-# we're building; this is intentional.
+# Download the delphix upgrade verification debian package, stored in the
+# combined-packages bundle.
 #
+AWS_S3_URI_COMBINED_PACKAGES=$(resolve_s3_uri \
+	"$AWS_S3_URI_COMBINED_PACKAGES" \
+	"devops-gate/master/linux-pkg/${UPSTREAM_BRANCH}/combine-packages/post-push/latest")
 
-AWS_S3_URI_UPGRADE_VERIFICATION=$(resolve_s3_uri \
-	"$AWS_S3_URI_UPGRADE_VERIFICATION" \
-	"$AWS_S3_PREFIX_UPGRADE_VERIFICATION" \
-	"devops-gate/master/upgrade-verify/master/post-push/release/latest")
+WORK_DIRECTORY=$(mktemp -d -p "$TOP/upgrade" tmp.pkgs.XXXXXXXXXX)
 
-download_delphix_s3_debs "$TOP/upgrade/debs" "$AWS_S3_URI_UPGRADE_VERIFICATION"
+download_combined_packages_artifacts "$AWS_S3_URI_COMBINED_PACKAGES" \
+	"$WORK_DIRECTORY" upgrade-verify
+
+extract_debs_into_dir "$WORK_DIRECTORY/packages/upgrade-verify" \
+	"$TOP/upgrade/debs"
+
+rm -rf "$WORK_DIRECTORY"
 
 #
 # Generate an Aptly/APT repository

--- a/scripts/build-ancillary-repository.sh
+++ b/scripts/build-ancillary-repository.sh
@@ -62,37 +62,6 @@ function build_ancillary_repository() {
 }
 
 #
-# The packages produced by Delphix are stored in Amazon S3.
-# Thus, in order to populate the ancillary repository with these
-# packages, they must be downloaded from S3, so they can be then
-# inserted into the Aptly repository.
-#
-# Here, we determine the URI of each of the Delphix packages, and
-# then use these URIs to download the packages later. Making this
-# determination is a little complex, and is dependent on the policy set
-# forth by the teams producing and storing the packages.
-#
-# With that said, there's three main methods of influencing the URI from
-# which the packages are downloaded:
-#
-# 1. If the package specific AWS_S3_URI environment variable is provided
-#    (e.g. AWS_S3_URI_UPGRADE_VERIFICATION), then this URI will be used to
-#    download the package. This is the simplest case, and enables the
-#    user of this script to directly influence this script.
-#
-# 2. If the package specific AWS_S3_PREFIX environment variable is
-#    provided (e.g. AWS_S3_PREFIX_UPGRADE_VERIFICATION), then this value is
-#    used to build the URI that will be used based on the default S3
-#    bucket that is used.
-#
-# 3. If nether the package specific AWS_S3_URI nor AWS_S3_PREFIX
-#    variables are provided, then logic kicks in to attempt to
-#    dynamically determine the URI of the most recently built package,
-#    and then uses that URI. This way, a naive user can not set any
-#    environment variables, and the script will work as expected.
-#
-
-#
 # Set UPSTREAM_BRANCH. This will determine which version of the linux package
 # mirror is used.
 #
@@ -115,32 +84,45 @@ else
 fi
 echo "Running with UPSTREAM_BRANCH set to ${UPSTREAM_BRANCH}"
 
+#
+# The packages produced by Delphix are stored in Amazon S3.
+# Thus, in order to populate the ancillary repository with these
+# packages, they must be downloaded from S3, so they can be then
+# inserted into the Aptly repository.
+#
+# All the Delphix-built packages consumed by appliance-build are compiled by
+# the combine-packages job. If a combine-packages URI is provided, fetch the
+# packages from there, otherwise determine the latest combined-packages URI
+# automatically.
+#
+
 AWS_S3_URI_COMBINED_PACKAGES=$(resolve_s3_uri \
-	"$AWS_S3_URI_COMBINED_PACKAGES" "" \
+	"$AWS_S3_URI_COMBINED_PACKAGES" \
 	"devops-gate/master/linux-pkg/${UPSTREAM_BRANCH}/combine-packages/post-push/latest")
 
-#
-# All package files will be placed into this temporary directory, such
-# that we can later point Aptly at this directory to build the Aptly/APT
-# repository.
-#
 mkdir -p "$TOP/build"
-PKG_DIRECTORY=$(mktemp -d -p "$TOP/build" tmp.pkgs.XXXXXXXXXX)
+WORK_DIRECTORY=$(mktemp -d -p "$TOP/build" tmp.pkgs.XXXXXXXXXX)
 
 #
-# Now that we've determined the URI of the Delphix-built packages, we can
-# download them.
+# Download all package artifacts built by Delphix, which includes debs and
+# metadata.
 #
-download_delphix_s3_debs_multidir "$PKG_DIRECTORY" "$AWS_S3_URI_COMBINED_PACKAGES/packages"
+mkdir -p "$WORK_DIRECTORY/artifacts"
+download_combined_packages_artifacts "$AWS_S3_URI_COMBINED_PACKAGES" \
+	"$WORK_DIRECTORY/artifacts"
 
 #
-# Now that our temporary package directory has been populated with all
-# first-party packages needed by live-build, we use this directory to
-# build up our Aptly/APT ancillary repository. After this function
+# Find all debs and put them into a directory that will be fed into Aptly.
+#
+mkdir -p "$WORK_DIRECTORY/debs"
+extract_debs_into_dir "$WORK_DIRECTORY/artifacts" "$WORK_DIRECTORY/debs"
+
+#
+# Build up our Aptly/APT ancillary repository. After this function
 # completes, there should be a directory named "ancillary-repository" at
 # the top level of the git repository, that can later be "aptly
 # serve"-ed and consumed by live-build.
 #
-build_ancillary_repository "$PKG_DIRECTORY"
+build_ancillary_repository "$WORK_DIRECTORY/debs"
 
-rm -rf "$PKG_DIRECTORY"
+rm -rf "$WORK_DIRECTORY"

--- a/scripts/common.sh
+++ b/scripts/common.sh
@@ -15,6 +15,7 @@
 # limitations under the License.
 #
 
+# shellcheck disable=SC2034
 TOP=$(git rev-parse --show-toplevel 2>/dev/null)
 
 function die() {
@@ -24,8 +25,7 @@ function die() {
 
 function resolve_s3_uri() {
 	local pkg_uri="$1"
-	local pkg_prefix="$2"
-	local latest_subprefix="$3"
+	local latest_subprefix="$2"
 
 	local bucket="snapshot-de-images"
 	local jenkinsid="jenkins-ops"
@@ -33,10 +33,6 @@ function resolve_s3_uri() {
 
 	if [[ -n "$pkg_uri" ]]; then
 		resolved_uri="$pkg_uri"
-	elif [[ "$pkg_prefix" == s3* ]]; then
-		resolved_uri="$pkg_prefix"
-	elif [[ -n "$pkg_prefix" ]]; then
-		resolved_uri="s3://$bucket/$pkg_prefix"
 	elif [[ -n "$latest_subprefix" ]]; then
 		aws s3 cp --quiet \
 			"s3://$bucket/builds/$jenkinsid/$latest_subprefix" .
@@ -55,40 +51,87 @@ function resolve_s3_uri() {
 	fi
 }
 
-function download_delphix_s3_debs() {
-	local pkg_directory="$1"
-	local S3_URI="$2"
-	local tmp_directory
+#
+# Given an S3 URI pointing to combined-packages artifacts, download all of its
+# artifacts to target directory. If a package name is passed as an argument,
+# then only copy the artifacts for that package.
+#
+# When the combine-packages Jenkins job generates artifacts, it does not
+# copy around the artifacts for individual packages. Rather, it creates a
+# COMPONENTS file that has links to each package's artifacts.
+#
+# When appliance-build is ran via Jenkins, the Jenkins job copies the original
+# combined-packages artifacts to a new S3 location, then dereferences the
+# COMPONENTS file and copies all individual package artifacts into a
+# "packages" directory created under that new S3 location. Jenkins then passes
+# that combined-packages URI to live-build.
+#
+# Thus if a "packages" directory is found under the combined-packages S3 URI,
+# we assume that the dereferencing has already been done and so we just sync
+# the whole directory. Otherwise, we must dereference the COMPONENTS file here
+# and fetch the artifacts for each package.
+#
+# Here are the files that are expected to be found after the download.
+# <combined packages base directory>/
+#   COMPONENTS
+#   ... (some other metadata files)
+#   packages/
+#     package1/
+#       ... (package 1 artifacts)
+#     package2/
+#       ... (package 2 artifacts)
+#     ... (remaining packages' artifacts)
+#
+# shellcheck disable=SC2164
+function download_combined_packages_artifacts() {
+	local combined_pkgs_uri="$1"
+	local target_dir="$2"
+	local pkg="$3"
 
-	tmp_directory=$(mktemp -d -p "$TOP/build" tmp.s3-debs.XXXXXXXXXX)
-	pushd "$tmp_directory" &>/dev/null
+	pushd "$target_dir" &>/dev/null
 
-	aws s3 sync --only-show-errors "$S3_URI" .
-	sha256sum -c --strict SHA256SUMS
+	if [[ -n "$pkg" ]]; then
+		aws s3 sync --exclude 'packages/*' --include "packages/$pkg/*" \
+			--only-show-errors "$combined_pkgs_uri" .
+	else
+		aws s3 sync --only-show-errors "$combined_pkgs_uri" .
+	fi
 
-	mv ./*deb "$pkg_directory/"
+	if [[ -d packages ]]; then
+		popd &>/dev/null
+		return
+	fi
+
+	[[ -f COMPONENTS ]] || die "COMPONENTS file missing."
+	mkdir packages
+	pushd packages &>/dev/null
+
+	local pkgname s3uri
+	while read -r line; do
+		pkgname=$(echo "$line" | cut -d: -f 1 | tr -d '[:space:]')
+		s3uri=$(echo "$line" | cut -d: -f 2- | tr -d '[:space:]')
+		[[ -n "$pkg" ]] && [[ "$pkg" != "$pkgname" ]] && continue
+		mkdir "$pkgname"
+		pushd "$pkgname" &>/dev/null
+		aws s3 sync --only-show-errors "$s3uri" .
+		sha256sum -c --strict SHA256SUMS
+		popd &>/dev/null
+	done <../COMPONENTS
 
 	popd &>/dev/null
-	rm -rf "$tmp_directory"
+	popd &>/dev/null
 }
 
-function download_delphix_s3_debs_multidir() {
-	local pkg_directory="$1"
-	local S3_URI="$2"
-	local tmp_directory
+#
+# Find all .deb and .ddeb packages in source directory tree and move them
+# to target directory.
+#
+function extract_debs_into_dir() {
+	local source_dir="$1"
+	local target_dir="$2"
 
-	tmp_directory=$(mktemp -d -p "$TOP/build" tmp.s3-debs.XXXXXXXXXX)
-	pushd "$tmp_directory" &>/dev/null
-
-	aws s3 sync --only-show-errors "$S3_URI" .
-
-	for subdir in */; do
-		pushd "$subdir" &>/dev/null
-		sha256sum -c --strict SHA256SUMS
-		mv ./*deb "$pkg_directory/"
-		popd &>/dev/null
-	done
-
-	popd &>/dev/null
-	rm -rf "$tmp_directory"
+	[[ -d "$target_dir" ]] ||
+		die "'$target_dir' must be an existing directory"
+	find "$source_dir" -name '*.deb' -exec mv {} "$target_dir" \;
+	find "$source_dir" -name '*.ddeb' -exec mv {} "$target_dir" \;
 }

